lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -1120,15 +1124,15 @@ def private_key(self) -> Optional[PrivateKey]:
         private_key = secret.get_content(refresh=True)["private-key"]
         return PrivateKey.from_string(private_key)
 
-    def _generate_private_key(self) -> None:
+    def _ensure_private_key(self) -> None:
         if self._private_key_generated():
             return
-        private_key = generate_private_key()
+        private_key = self._private_key or generate_private_key()
```

Comment:
Does it make sense to put it in a secret? This would mean that we'd have 2 sources for the private key: the input + the secret.

Reply:
I thought about it. I think we should have the secret be the source of truth and only set it to the user provided secret the first time and completely manage the key from there.
- We make sure not to override the key since we check whether a key was generated before storing the user provided key.
- This will allow rotation as we will override the user provided key and not lose that state because of the same guard mentioned in the previous point.

I think for the migration it makes sense to allow the users to initialise the class with their own key, but we take care of the rest. 

One further enhancement that I added is allowing the same logic in the `regenerate_private_key`